### ISSUE TITLE
False positive for implied_bounds_in_impls

### ISSUE THREAD

**Issue by e00E** [rust-lang/rust-clippy#12576]
### Summary

Clippy warns that in a function like `fn foo() -> impl Iterator<Item = ()> + FusedIterator`  the Iterator bound is superfluous. This is wrong. The bound is needed to specify the iterator's item type. Clippy's suggestion to remove the Iterator bound doesn't compile.

### Lint Name

implied_bounds_in_impls

### Reproducer

I tried this code:

```rust
fn foo() -> impl Iterator<Item = ()> + std::iter::FusedIterator {
    std::iter::empty()
}

fn bar() {
    for _ in foo() {}
}
```

I saw this happen:

```
warning: this bound is already specified as the supertrait of `std::iter::FusedIterator`
 --> src/lib.rs:1:18
  |
1 | fn foo() -> impl Iterator<Item = ()> + std::iter::FusedIterator {
  |                  ^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#implied_bounds_in_impls
  = note: `#[warn(clippy::implied_bounds_in_impls)]` on by default
help: try removing this bound
  |
1 - fn foo() -> impl Iterator<Item = ()> + std::iter::FusedIterator {
1 + fn foo() -> impl std::iter::FusedIterator<Item = ()> {
```

I expected to see this happen:

no warning

### Version

```text
rustc 1.77.0 (aedd173a2 2024-03-17)
binary: rustc
commit-hash: aedd173a2c086e558c2b66d3743b344f977621a7
commit-date: 2024-03-17
host: x86_64-unknown-linux-gnu
release: 1.77.0
LLVM version: 17.0.6
```

**Comment by y21** [comment]
> Clippy's suggestion to remove the Iterator bound doesn't compile.

clippy's suggestion seem to compile just fine: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4a3501a2e66985e1692037e7462428a8

I don't see why this would be wrong. The associated type `Item` is included on the implying bound in the suggestion.

**Comment by e00E** [comment]
You're right. I misread the message as telling me to remove the Iterator bound without moving the `Item=` over. I didn't know you *could* move it over.

**Comment by y21** [comment]
Although to be fair, the lint also doesn't make it super clear that moving associated types might also be needed if you aren't reading the suggestion carefully, so I can see how that can easily be missed.
Maybe it would help if the `help:` message was modified to also say something like "... and move the necessary associated type bindings to the subtrait"? 🤔

**Comment by e00E** [comment]
It would help in the case where there are associated types. It could be confusing in the case where they aren't. If clippy can distinguish the cases, then it's a pure improvement. The suggestion right below does tell you exactly what to do so it's not a big deal to keep it the way it is.

**Comment by SimonSapin** [comment]
> Maybe it would help if the `help:` message was modified to also say something like "... and move the necessary associated type bindings to the subtrait"? 🤔

Yes this would help. I came here to file what I thought was a false positive because it seemed like Clippy’s suggestion was leaving nowhere to specify the associated type. I didn’t realize `impl ExactSizeIterator<Item = char>` was legal (with the associated type being on a super-trait, not on `ExactSizeIterator` itself)
